### PR TITLE
rsa_signverify_init: Set the PARAMS after key is set

### LINKED ISSUE #17075: Segfault in EVP_PKEY_sign_init_ex with RSA_PKCS1_PSS_PADDING
Hi all,

I'm trying to use the low-level primitives (EVP_PKEY_sign_init_ex) in OpenSSL to create a PSS signature. However, I get a segmentation fault from OpenSSL. I'm not sure if I'm doing something wrong, or if this is a bug (as such, I made this a question).
The following code reproduces the error:
```
#include <openssl/core_names.h>
#include <openssl/evp.h>
#include <openssl/param_build.h>
#include <openssl/params.h>
#include <openssl/rsa.h>

EVP_PKEY* generateKey() {
  EVP_PKEY*     pkey = NULL;
  EVP_PKEY_CTX* ctx  = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
  EVP_PKEY_keygen_init(ctx);
  EVP_PKEY_generate(ctx, &pkey);
  assert(pkey);
  return pkey;
}

OSSL_PARAM* generateSignParams() {
  OSSL_PARAM_BLD* bld = OSSL_PARAM_BLD_new();
  OSSL_PARAM_BLD_push_int(bld, OSSL_PKEY_PARAM_PAD_MODE, RSA_PKCS1_PSS_PADDING);
  // This does not seem to have any effect:
  // OSSL_PARAM_BLD_push_utf8_string(bld, OSSL_SIGNATURE_PARAM_DIGEST, "sha1", 4);

  OSSL_PARAM* result = OSSL_PARAM_BLD_to_param(bld);
  assert(result);
  return result;
}

int main() {
  EVP_PKEY*     key = generateKey();
  EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);

  // This does not seem to have any effect:
  // EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha1());

  EVP_PKEY_sign_init_ex(ctx, generateSignParams());  // <----------- crash in this line
}
```

This crashes within `EVP_PKEY_sign_init_ex`. The stack trace at the point of the crash is as follows:
```
test.exe!EVP_MD_is_a(const evp_md_st * md, const char * name) Line 716
test.exe!rsa_check_padding(const PROV_RSA_CTX * prsactx, const char * mdname, const char * mgf1_mdname, int mdnid) Line 143
test.exe!rsa_setup_md(PROV_RSA_CTX * ctx, const char * mdname, const char * mdprops) Line 297
test.exe!rsa_set_ctx_params(void * vprsactx, const ossl_param_st * params) Line 1350
test.exe!rsa_signverify_init(void * vprsactx, void * vrsa, const ossl_param_st * params, int operation) Line 404
test.exe!rsa_sign_init(void * vprsactx, void * vrsa, const ossl_param_st * params) Line 504
test.exe!evp_pkey_signature_init(evp_pkey_ctx_st * ctx, int operation, const ossl_param_st * params) Line 473
test.exe!EVP_PKEY_sign_init_ex(evp_pkey_ctx_st * ctx, const ossl_param_st * params) Line 563
```

The `md` parameter to `EVP_MD_is_a` is `NULL` (which is not allowed). This value is retrieved from `prsactx->md`, in the following code (`rsa_check_padding`):
```
                if ((mdname != NULL && !EVP_MD_is_a(prsactx->md, mdname))
                    || (mgf1_mdname != NULL
                        && !EVP_MD_is_a(prsactx->mgf1_md, mgf1_mdname))) {
```
Here, `mdname` is not `NULL`, as it was given a default value in `rsa_set_ctx_params`, due to the following code:
```
    if (prsactx->md == NULL && pmdname == NULL
        && pad_mode == RSA_PKCS1_PSS_PADDING)
        pmdname = RSA_DEFAULT_DIGEST_NAME;
```

So I can understand the crash, but I'm not sure how to fix it. Setting the message digest (as commented in my reproduction code) does not seem to have any effect.

Am I doing something wrong here, or is this a bug? How can I get this to work?

I know that there may be higher level primitives that are more suitable for this specific scenario, but we will sometimes need lower level primitives, so it would be confusing if that doesn't always work as expected.

Thanks in advance!

Best regards,
SpoofedEx

### PR DESCRIPTION
**Pull request by t8m**
Also, default to unrestricted pss parameters until the key is set.

Fixes #17075

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: providers/implementations/signature/rsa_sig.c [hunk 1]
```diff
@@ -406,9 +409,6 @@ static int rsa_signverify_init(void *vprsactx, void *vrsa,
 
     prsactx->operation = operation;
 
-    if (!rsa_set_ctx_params(prsactx, params))
-        return 0;
-
     /* Maximum for sign, auto for verify */
     prsactx->saltlen = RSA_PSS_SALTLEN_AUTO;
     prsactx->min_saltlen = -1;
```
Comment (mattcaswell) [inline providers/implementations/signature/rsa_sig.c:414]: Why do we need this block in *both* the newctx function and here?

**Comment by mattcaswell** [comment]
CI failures look relevant

File: providers/implementations/signature/rsa_sig.c (hunk 1 above)
Comment (t8m) [reply providers/implementations/signature/rsa_sig.c:414]: If we call the sign_init multiple times, it could be needed.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for the review.

**Pull request closed without merging**